# Red Neuronal (Keras)

<center>

__Modelación del Aprendizaje con Inteligencia Artificial__

_GPO 302_

Actividad 3 _Redes Neuronales (Keras)_


Prof. Jorge Adolfo Ramírez Uresti




Diego A. Parra Ruíz - - - A01801026
</center>

## Importaciones

In [ ]:
# first neural network with keras tutorial
from numpy import loadtxt
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras import backend as K
import tensorflow as tf
import sklearn.metrics

## Enlace a Dataset

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# load the dataset
# Drive File: dataset = loadtxt('/content/drive/MyDrive/Urestia/indiabetes.csv', delimiter=',')
dataset = loadtxt('/Data/Diabetes.csv', delimiter=',')
# split into input (X) and output (y) variables
X = dataset[:,0:8]
y = dataset[:,8]

In [ ]:
def keras_f1_score(y_true, y_pred):
    # Cast to float32 for compatibility
    y_true = K.cast(y_true, 'float32')
    # Round predictions to get binary output (0 or 1)
    y_pred = K.cast(K.round(y_pred), 'float32')

    # Calculate True Positives, False Positives, and False Negatives using Keras backend operations
    # TP: y_true is 1 and y_pred is 1
    true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))
    # PP: y_true is 1
    possible_positives = K.sum(K.round(K.clip(y_true, 0, 1)))
    # PR: y_pred is 1
    predicted_positives = K.sum(K.round(K.clip(y_pred, 0, 1)))

    # Calculate Precision and Recall
    # Use K.epsilon() to avoid division by zero
    precision = true_positives / (predicted_positives + K.epsilon())
    recall = true_positives / (possible_positives + K.epsilon())

    # Calculate F1 Score
    # Use K.epsilon() to avoid division by zero
    f1_score = 2 * ((precision * recall) / (precision + recall + K.epsilon()))

    return f1_score

## Modelo Original

### Creación de Clase Secuencial

In [ ]:
# define the keras model
model = Sequential()
model.add(Dense(12, input_shape=(8,), activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

### Compilación del Modelo

In [ ]:
# compile the keras model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

### Activación del Modelo

In [ ]:
# fit the keras model on the dataset
model.fit(X, y, epochs=150, batch_size=10)

Epoch 1/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 2s 10ms/step - accuracy: 0.5191 - loss: 7.5890
Epoch 2/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.5631 - loss: 2.4172
Epoch 3/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.5771 - loss: 1.7376
Epoch 4/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.5806 - loss: 1.4393
Epoch 5/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6099 - loss: 1.2785
Epoch 6/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6304 - loss: 1.2243
Epoch 7/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6451 - loss: 0.9048
Epoch 8/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6281 - loss: 0.9899
Epoch 9/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6666 - loss: 0.9305
Epoch 10/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6085 - loss: 1.0325
Epoch 11/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6582 - loss: 0.9191
Epoch 12/150
77/77 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy

### Evaluación del Modelo

In [ ]:
# evaluate the keras model
_, accuracy = model.evaluate(X, y) #Discards the loss with a garbage "_" when generated
print('Accuracy: %.2f' % (accuracy*100))

24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7436 - loss: 0.4958  
Accuracy: 77.34


### Predicción del Modelo

In [ ]:
# make class predictions with the model
predictions = (model.predict(X) > 0.5).astype(int)
# summarize the first 5 cases
for i in range(5):
	print('%s => %d (expected %d)' % (X[i].tolist(), predictions[i], y[i]))

24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step  
[6.0, 148.0, 72.0, 35.0, 0.0, 33.6, 0.627, 50.0] => 1 (expected 1)
[1.0, 85.0, 66.0, 29.0, 0.0, 26.6, 0.351, 31.0] => 0 (expected 0)
[8.0, 183.0, 64.0, 0.0, 0.0, 23.3, 0.672, 32.0] => 1 (expected 1)
[1.0, 89.0, 66.0, 23.0, 94.0, 28.1, 0.167, 21.0] => 0 (expected 0)
[0.0, 137.0, 40.0, 35.0, 168.0, 43.1, 2.288, 33.0] => 1 (expected 1)


<ipython-input-62-144830af59af>:5: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)
  print('%s => %d (expected %d)' % (X[i].tolist(), predictions[i], y[i]))


In [ ]:
f1 = sklearn.metrics.f1_score(y.astype(int), predictions.flatten())
print('Model F1 Score: %.2f' % (f1*100), '%')

Model F1 Score: 62.66 %


## Modelo Modificado

### Nueva Capa Neuronal

In [ ]:
# define the keras model
model = Sequential()
model.add(Dense(12, input_shape=(8,), activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(4, activation='relu')) # Se agrega nueva Capa de 4 neuronas al modelo
model.add(Dense(1, activation='sigmoid'))

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


### Compilación del Modelo

In [ ]:
# compile the keras model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

# fit the keras model on the dataset
model.fit(X, y, epochs=120, batch_size=12) # Se cambia el learning a 180 ciclos del dset y Delta_peso cada 16 datapoints

Epoch 1/120
64/64 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.6120 - loss: 16.4491
Epoch 2/120
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5838 - loss: 4.1189
Epoch 3/120
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5536 - loss: 3.1570
Epoch 4/120
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6093 - loss: 1.9769
Epoch 5/120
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6316 - loss: 1.4527
Epoch 6/120
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5965 - loss: 1.3327
Epoch 7/120
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5754 - loss: 1.4000
Epoch 8/120
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5903 - loss: 1.1540
Epoch 9/120
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5893 - loss: 1.0897
Epoch 10/120
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6162 - loss: 0.9119
Epoch 11/120
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6284 - loss: 0.9080
Epoch 12/120
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy

### Evaluación y Predicción del Modelo

In [ ]:
# evaluate the keras model
_, accuracy = model.evaluate(X, y, verbose=0) # Evaluate again to get the Keras F1 score if needed, or simply avoid overwriting in the previous cell
print('Accuracy: %.2f' % (accuracy*100))

Accuracy: 76.95


In [ ]:
predictions = (model.predict(X) > 0.5).astype(int)
f1 = sklearn.metrics.f1_score(y.astype(int), predictions.flatten())
print('Model F1 Score: %.2f' % (f1*100), '%')

24/24 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step  
Model F1 Score: 66.29 %


## Comparación de Modelos

Es posible observar a través de la comparación de métricas de los dos modelos cuentan con una medida de Accuracy similar, siempre rondando alrededor del 77%.

Asimismo, al considerar como evaluar el dataset, se considero que el modelo con F1_Score sería más adecuado. Esto dado que en el contexto de la diabetes es conveniente tener una métrica balanceada que tome en cuenta tanto los falsos positivos como los falsos negativos, dado que es importante intentar clasificar al máximo las personas que cuenten con diabtes para brindar un buen tratamiento, pero también es importante observar que otro subconjunto no sea clasificado erroneamente y sufra las consecuencias de un tratamiento a un padecimiento con el que no cuenta. Es decir, que es conveniente mantener un balance entre estas partes de la matriz de confusión y por ello se eligió el promedio ponderado generado por el F1 Score.

En este caso, las métricas de ambos modelos, el de 3 capas y el de 4 capas son bastante similares, teniendo solo un diferenciador significativo en el área del F1_Score, dado que hay en promedio cada ocasión que es corrido el programa un cambio de 8% aproximadamente entre el primer y segundo modelo; teniendo el primer modelo un 62% y el segundo un 66% en la última corrida. Lo que indica que el segundo modelo es un poco mejor que el primero para este ámbito y es algo mejor elegirle para el fin de clasificación diabético que se presenta, sin embargo tomando en cuenta que su F1_Score no es lo suficientemente alto como para coodepender totalmente del modelo en la diagnosis.